# Page Visits Funnel

The goal of this project is to analyze data on visits to a website of Cool T-Shirts Inc. My job, as an analyst is to build a funnel, which is a description of how many people continue to the next step of a multi-step process, using Pandas.

In this case, our funnel is going to describe the following process:

    1. A user visits CoolTShirts.com
    2. A user adds a t-shirt to their cart
    3. A user clicks “checkout”
    4. A user actually purchases a t-shirt

First we need to import Pandas library and prepare csv files that we are going to analyze. Files contain the fillowing info:

- Visits file - list of all users who have visited the website.
- Cart file - list of all users who have added a t-shirt to their cart.
- Checkout - list of all users who have started the checkout.
- Purchase - list of all users who have purchased a t-shirt.


In [1]:
import pandas as pd

# Read csv files 
visits = pd.read_csv('visits.csv',
                     parse_dates=[1])
cart = pd.read_csv('cart.csv',
                   parse_dates=[1])
checkout = pd.read_csv('checkout.csv',
                       parse_dates=[1])
purchase = pd.read_csv('purchase.csv',
                       parse_dates=[1])


Let's inspect the DataFrames to know exactly what we are dealing with. Below is an example of the first 5 rows of the each given csv file in piles.

In [2]:
pd.concat([visits.head(5), cart.head(5), checkout.head(5), purchase.head(5)])

user_id          visit_time  \
0  943647ef-3682-4750-a2e1-918ba6f16188 2017-04-07 15:14:00   
1  0c3a3dd0-fb64-4eac-bf84-ba069ce409f2 2017-01-26 14:24:00   
2  6e0b2d60-4027-4d9a-babd-0e7d40859fb1 2017-08-20 08:23:00   
3  6879527e-c5a6-4d14-b2da-50b85212b0ab 2017-11-04 18:15:00   
4  a84327ff-5daa-4ba1-b789-d5b4caf81e96 2017-02-27 11:25:00   
0  2be90e7c-9cca-44e0-bcc5-124b945ff168                 NaT   
1  4397f73f-1da3-4ab3-91af-762792e25973                 NaT   
2  a9db3d4b-0a0a-4398-a55a-ebb2c7adf663                 NaT   
3  b594862a-36c5-47d5-b818-6e9512b939b3                 NaT   
4  a68a16e2-94f0-4ce8-8ce3-784af0bbb974                 NaT   
0  d33bdc47-4afa-45bc-b4e4-dbe948e34c0d                 NaT   
1  4ac186f0-9954-4fea-8a27-c081e428e34e                 NaT   
2  3c9c78a7-124a-4b77-8d2e-e1926e011e7d                 NaT   
3  89fe330a-8966-4756-8f7c-3bdbcd47279a                 NaT   
4  3ccdaf69-2d30-40de-b083-51372881aedd                 NaT   
0  4b44ace4-2721-47a0-b24b-15fbfa2abf85                 NaT   
1  02e684ae-a448-408f-a9ff-dcb4a5c99aac                 NaT   
2  4b4bc391-749e-4b90-ab8f-4f6e3c84d6dc                 NaT   
3  a5dbb25f-3c36-4103-9030-9f7c6241cd8d                 NaT   
4  46a3186d-7f5a-4ab9-87af-84d05bfd4867                 NaT   

            cart_time       checkout_time       purchase_time  
0                 NaT                 NaT                 NaT  
1                 NaT                 NaT                 NaT  
2                 NaT                 NaT                 NaT  
3                 NaT                 NaT                 NaT  
4                 NaT                 NaT                 NaT  
0 2017-11-07 20:45:00                 NaT                 NaT  
1 2017-05-27 01:35:00                 NaT                 NaT  
2 2017-03-04 10:38:00                 NaT                 NaT  
3 2017-09-27 08:22:00                 NaT                 NaT  
4 2017-07-26 15:48:00                 NaT                 NaT  
0                 NaT 2017-06-25 09:29:00                 NaT  
1                 NaT 2017-04-07 20:11:00                 NaT  
2                 NaT 2017-07-13 11:38:00                 NaT  
3                 NaT 2017-04-20 16:15:00                 NaT  
4                 NaT 2017-01-08 20:52:00                 NaT  
0                 NaT                 NaT 2017-05-11 04:25:00  
1                 NaT                 NaT 2017-09-05 08:45:00  
2                 NaT                 NaT 2017-11-20 20:49:00  
3                 NaT                 NaT 2017-01-22 15:18:00  
4                 NaT                 NaT 2017-06-11 11:32:00

What percent of users who visited Cool T-Shirts Inc. ended up not placing a t-shirt in their cart?

In [3]:
def to_perc(df1, df2, column_name):
  merge = pd.merge(df1, df2, how = 'left')
  print('{:.2%}'.format(merge[column_name].isna().sum() / len(merge)))
  return

# Get percentage of users who didn't place t-shirt in their cart
to_perc(visits, cart, 'cart_time')

82.60%


What percentage of users put items in their cart, but did not proceed to checkout?

In [4]:
# Get percentage of users who put items in their cart, but did not proceed to checkout
to_perc(cart, checkout, 'checkout_time')

25.31%


What percentage of users proceeded to checkout, but did not purchase a t-shirt?

In [5]:
# Get percentage of users proceeded to checkout, but did not purchase a t-shirt
to_perc(checkout, purchase, 'purchase_time')

16.89%


What is the average time from initial visit to final purchase?

In [6]:
# Average time from initial visit to final purchase
a = pd.merge(visits, purchase, how='right')
a['time_to_purchase'] = a.purchase_time - a.visit_time
print(a.time_to_purchase.mean())

0 days 00:43:12.380952380


### Conclusion 

If in our analysis we consider only one factor - the highest percentage of users not completing a given step - then the first step of the funnel would be the weakest. It has `82.60%` of those who didn't even add anything in the cart. It can indicate that products look not enticing enough by themselves or are represented poorly on the website, or maybe the process of adding a product to a cart is not clear for users. 

So the answer to the main question: how might Cool T-Shirts Inc. change their website to fix this problem? - might consist of the following steps:
- Analyse how users find the goods personally? Do they find what they seek? 
- Analyse goods' representation on the website. Is design optimal, goal-oriented and esthetically pleasing? 
- Analyse the process of adding a good to a cart. Does it trigger focused thinking processes or the actions can be performed in an easy, nearly automatic way? 

Depending on the answers to these questions Cool T-Shirts Inc. would have to either redesign their T-shots or bits of their website. 

